### PR TITLE
scheduler: enable secure port and authn/z

### PR DESCRIPTION
**Pull request by sttts**
Completes https://github.com/kubernetes/kubernetes/pull/67192.

This PR adds a secure port 10259 to the scheduler
- enabled by default as for the controller manager
- deprecates the old insecure port
- adds exhaustive delegated authn/z tests.

```release-note
Add secure port 10259 to the kube-scheduler (enabled by default) and deprecate old insecure port 10251. Without further flags self-signed certs are created on startup in memory.
```

Fixes https://github.com/kubernetes/kubernetes/issues/58983.

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@sttts thanks, will review it soon.

File: cmd/kube-scheduler/app/server.go [hunk 1]
```diff
@@ -130,22 +130,28 @@ func run(cmd *cobra.Command, args []string, opts *options.Options) error {
 	// To help debugging, immediately log version
 	glog.Infof("Version: %+v", version.Get())
 
-	// Apply algorithms based on feature gates.
-	// TODO: make configurable?
-	algorithmprovider.ApplyFeatureGates()
```
Comment (aveshagarwal) [inline cmd/kube-scheduler/app/server.go]: why is this removed?

Comment (aveshagarwal) [reply cmd/kube-scheduler/app/server.go]: oh nevermind, I see it below in Run().

File: cmd/kube-scheduler/app/options/insecure_serving.go [hunk 2]
```diff
@@ -31,8 +31,8 @@ import (
 // CombinedInsecureServingOptions sets up to two insecure listeners for healthz and metrics. The flags
 // override the ComponentConfig and DeprecatedInsecureServingOptions values for both.
 type CombinedInsecureServingOptions struct {
-	Healthz *apiserveroptions.DeprecatedInsecureServingOptions
-	Metrics *apiserveroptions.DeprecatedInsecureServingOptions
+	Healthz *apiserveroptions.DeprecatedInsecureServingOptionsWithLoopback
+	Metrics *apiserveroptions.DeprecatedInsecureServingOptionsWithLoopback
```
[the remark this answers is not in the document]
Comment (aveshagarwal) [reply cmd/kube-scheduler/app/options/insecure_serving.go]: it seems that is it extending it to loopback? though still not sure why to extend a deprecated option?

[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kube-scheduler/app/options/insecure_serving.go]: It's not extending it in the code sense. We have support for insecure ports using these two option structs (not new for scheduler). The whole insecure port code is deprecated, to be removed when all components switched over to secure ports. The difference of `DeprecatedInsecureServingOptionsWithLoopback` is just that it add the loopback authenticator which helps testing. Otherwise it is the same.

**Comment by aveshagarwal** [comment]
@sttts overall lgtm, see if @k82cn and @bsalamat have more feedback?

**Comment by sttts** [comment]
@aveshagarwal  ptal

/assign @tossmilestone @bsalamat

File: cmd/kube-scheduler/app/options/options.go [hunk 3]
```diff
@@ -85,25 +86,33 @@ func NewOptions() (*Options, error) {
...
 			UseLegacyPolicyConfig:    false,
 			PolicyConfigMapNamespace: metav1.NamespaceSystem,
 		},
 	}
 
+	o.Authentication.RemoteKubeConfigFileOptional = true
+	o.Authorization.RemoteKubeConfigFileOptional = true
+	o.Authorization.AlwaysAllowPaths = []string{"/healthz"}
+
+	o.SecureServing.ServerCert.CertDirectory = "/var/run/kubernetes"
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options.go]: set to empty by default to generate in-memory?

File: test/e2e/network/firewall.go [hunk 4]
```diff
@@ -174,15 +174,17 @@ var _ = SIGDescribe("Firewall rule", func() {
...
+		flag, _ := framework.TestNotReachableHTTPTimeout(masterAddr, ports.InsecureKubeControllerManagerPort, gce.FirewallTestTcpTimeout, false)
 		Expect(flag).To(BeTrue())
-		flag, _ = framework.TestNotReachableHTTPTimeout(masterAddr, ports.SchedulerPort, gce.FirewallTestTcpTimeout)
+		flag, _ = framework.TestNotReachableHTTPTimeout(masterAddr, ports.KubeSchedulerPort, gce.FirewallTestTcpTimeout, true)
 		Expect(flag).To(BeTrue())
-		flag, _ = framework.TestNotReachableHTTPTimeout(nodeAddrs[0], ports.KubeletPort, gce.FirewallTestTcpTimeout)
+		flag, _ = framework.TestNotReachableHTTPTimeout(masterAddr, ports.InsecureSchedulerPort, gce.FirewallTestTcpTimeout, false)
 		Expect(flag).To(BeTrue())
-		flag, _ = framework.TestNotReachableHTTPTimeout(nodeAddrs[0], ports.KubeletReadOnlyPort, gce.FirewallTestTcpTimeout)
+		flag, _ = framework.TestNotReachableHTTPTimeout(nodeAddrs[0], ports.KubeletPort, gce.FirewallTestTcpTimeout, false)
```
Comment (liggitt) [inline test/e2e/network/firewall.go]: surprised `KubeletPort` wasn't secure

Comment (tossmilestone) [reply test/e2e/network/firewall.go]: Further work should be done to secure the `kubelet`, we won't give it up 😆

File: cmd/kube-scheduler/app/options/options.go (hunk 3 above)
Comment (sttts) [reply cmd/kube-scheduler/app/options/options.go]: right, your PR merged.

Comment (sttts) [reply cmd/kube-scheduler/app/options/options.go]: done

**Comment by sttts** [comment]
/retest

This might be a flake that also shows up in other (innocent looking) PRs.

**Comment by spiffxp** [comment]
The pull-kubernetes-e2e-gke job is triggering because it will trigger for any PR that touches a file matching the regex `^(cluster/gce|cluster/addons).*$`.  This PR happens to touch `cluster/gce/gci/configure-helper.sh`.  Whether the failure is valid or not, I'm unsure.  Looking.

Going forward, the triggering regex for pull-kubernetes-e2e-gke probably only needs to cover `cluster/addons` at best, as we no longer have any `cluster/gke` scripts that source files from `cluster/gce`.  Going to file a PR to trim the regex down, FYI @roberthbailey @BenTheElder

**Comment by spiffxp** [comment]
The actual failure appears to be
```
W1107 09:45:44.064] ERROR: (gcloud.container.clusters.create) Operation [<Operation
W1107 09:45:44.065]  clusterConditions: []
W1107 09:45:44.065]  detail: u'All cluster resources were brought up, but the cluster API is reporting that: component "kube-apiserver" from endpoint "gke-d97f5d2405088e02923d-85eb" is unhealthy.'
W1107 09:45:44.065]  endTime: u'2018-11-07T09:45:42.630090979Z'
W1107 09:45:44.065]  name: u'operation-1541583255239-a1541e17'
W1107 09:45:44.065]  nodepoolConditions: []
W1107 09:45:44.065]  operationType: OperationTypeValueValuesEnum(CREATE_CLUSTER, 1)
W1107 09:45:44.065]  selfLink: u'https://test-container.sandbox.googleapis.com/v1/projects/943316867058/zones/us-central1-f/operations/operation-1541583255239-a1541e17'
W1107 09:45:44.065]  startTime: u'2018-11-07T09:34:15.239200848Z'
W1107 09:45:44.065]  status: StatusValueValuesEnum(DONE, 3)
W1107 09:45:44.066]  statusMessage: u'All cluster resources were brought up, but the cluster API is reporting that: component "kube-apiserver" from endpoint "gke-d97f5d2405088e02923d-85eb" is unhealthy.'
W1107 09:45:44.066]  targetLink: u'https://test-container.sandbox.googleapis.com/v1/projects/943316867058/zones/us-central1-f/clusters/e2e-3153-218b6'
W1107 09:45:44.066]  zone: u'us-central1-f'>] finished with error: All cluster resources were brought up, but the cluster API is reporting that: component "kube-apiserver" from endpoint "gke-d97f5d2405088e02923d-85eb" is unhealthy.
```

**Comment by spiffxp** [comment]
/retest

At a glance it doesn't look like an environment issue, but I'm not sure where to go for logs that would further indicate what's happening here
- tesgrid for presubmits: https://testgrid.k8s.io/presubmits-kubernetes-nonblocking#pull-kubernetes-e2e-gke
- triage for all CI & PR gke jobs: https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=gke

**Comment by sttts** [comment]
@spiffxp nothing visible in the kube-apiserver or kube-scheduler logs?

**Comment by sttts** [comment]
@liggitt can you take another look here? I took out the cluster/gce changes which for some reason break gke.

File: cmd/kube-scheduler/app/options/options.go [hunk 5]
```diff
@@ -200,6 +216,12 @@ func (o *Options) Validate() []error {
 
 // Config return a scheduler config object
 func (o *Options) Config() (*schedulerappconfig.Config, error) {
+	if o.SecureServing != nil {
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options.go]: no nil check needed here, MaybeDefaultWithSelfSignedCerts no-ops if SecureServing is nil

**Comment by liggitt** [comment]
one nit, up to you if you want to address it

/lgtm
/approve

**Comment by liggitt** [comment]
if you have the problematic gce changes in a commit, can you throw those in a new PR for inspection?

**Comment by sttts** [comment]
@liggitt done: https://github.com/kubernetes/kubernetes/pull/70800

File: cmd/kube-scheduler/app/options/options.go (hunk 5 above)
Comment (sttts) [reply cmd/kube-scheduler/app/options/options.go]: fixed

Comment (liggitt) [reply cmd/kube-scheduler/app/options/options.go]: I appear to have led you astray... sorry:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b607db]

goroutine 86 [running]:
testing.tRunner.func1(0xc00030b200)
	GOROOT/src/testing/testing.go:792 +0x6a7
panic(0x1d2c320, 0x2fd61e0)
	GOROOT/src/runtime/panic.go:513 +0x1b9
k8s.io/kubernetes/cmd/kube-scheduler/app/options.(*Options).Config(0xc0001b5c80, 0xc000073e01, 0x41d0f8, 0x568f5b)
	cmd/kube-scheduler/app/options/options.go:219 +0xcb
```

**Comment by AishSundar** [comment]
/cc @mengqiy 

Adding @mengqiy who is the GKE oncall. @mengqiy this PR is failing on ```pull-kubernetes-e2e-gke``` presubmits. can you plz help @sttts get the right GLKE CI logs to investigate why. Thanks

**Comment by sttts** [comment]
> Adding @mengqiy who is the GKE oncall. @mengqiy this PR is failing on `pull-kubernetes-e2e-gke` presubmits. can you plz help @sttts get the right GLKE CI logs to investigate why. Thanks

Note that we moved the commit which breaks GKE into https://github.com/kubernetes/kubernetes/pull/70800#issuecomment-437026266

**Comment by furkanmustafa** [comment]
> deprecates the old insecure port

@sttts I want to learn about the motivations behind deprecating the plain http (insecure!?) port. Can you point me to any related discussions about this?
